tests/test_sample_rayservice_yamls.py

```diff
@@ -1,55 +1,338 @@
-    }
-    logger.info(image_dict)
-
-    # Build a test plan
-    logger.info("Build a test plan ...")
-    test_cases = unittest.TestSuite()
-    for index, new_cr in enumerate(sample_yaml_files):
-        logger.info('[TEST %d]: %s', index, new_cr['name'])
-        addEvent = RayServiceAddCREvent(new_cr['cr'], [rs], 90, NAMESPACE, new_cr['path'])
-        test_cases.addTest(GeneralTestCase('runtest', image_dict, addEvent))
-
-    # Execute all tests
-    runner = unittest.TextTestRunner()
-    test_result = runner.run(test_cases)
-
-    # Without this line, the exit code will always be 0.
-    assert test_result.wasSuccessful()
+NAMESPACE = 'default'
+DEFAULT_IMAGE_DICT = {
+    CONST.RAY_IMAGE_KEY: os.getenv('RAY_IMAGE', default='rayproject/ray:2.5.0'),
+    CONST.OPERATOR_IMAGE_KEY: os.getenv('OPERATOR_IMAGE', default='kuberay/operator:nightly'),
+}
+
+class RayServiceAddCREvent(CREvent):
+    """CREvent for RayService addition"""
+
+    def exec(self):
+        shell_subprocess_run(f"kubectl apply -n {self.namespace} -f {self.filepath}")
+
+    def wait(self):
+        """Wait for RayService to converge
+
+        Wait until:
+          (1) The number of head pods and worker pods are as expected.
+          (2) All head pods and worker pods are "Running".
+          (3) Service named "rayservice-sample-serve" presents
+        """
+
+        logger.info("Waiting for pods in ray service to be running...")
+        k8s_v1_api = K8S_CLUSTER_MANAGER.k8s_client_dict[CONST.K8S_V1_CLIENT_KEY]
+
+        start_time = time.time()
+        expected_head_pods = get_expected_head_pods(self.custom_resource_object)
+        expected_worker_pods = get_expected_worker_pods(self.custom_resource_object)
+        while time.time() - start_time < self.timeout:
+            headpods = k8s_v1_api.list_namespaced_pod(
+                namespace = self.namespace, label_selector='ray.io/node-type=head')
+            workerpods = k8s_v1_api.list_namespaced_pod(
+                namespace = self.namespace, label_selector='ray.io/node-type=worker')
+            head_services = k8s_v1_api.list_namespaced_service(
+                namespace = self.namespace, label_selector =
+                f"ray.io/serve={self.custom_resource_object['metadata']['name']}-serve")
+
+            if (len(head_services.items) == 1 and len(headpods.items) == expected_head_pods
+                    and len(workerpods.items) == expected_worker_pods
+                    and check_pod_running(headpods.items) and check_pod_running(workerpods.items)):
+                logger.info("--- RayServiceAddCREvent %s seconds ---", time.time() - start_time)
+                return
+
+            time.sleep(1)
+
+        logger.info(f"RayServiceAddCREvent wait() failed to converge in {self.timeout}s.")
+        logger.info(
+            f"expected_head_pods: {expected_head_pods}, "
+            f"expected_worker_pods: {expected_worker_pods}"
+        )
+        show_cluster_info(self.namespace)
```

Comment:
Does this log the number of running head pods and worker pods? If not, could you log those?

Reply:
Logged!